Revamp Service Bindings docs for RPC (#13252)
cloudflare/cloudflare-docs: content/workers/runtime-apis/bindings/service-bindings.md

## About Service bindings

[Service bindings](/workers/configuration/bindings/about-service-bindings/) facilitate Worker-to-Worker communication. A Service binding allows Worker A to call a method on Worker B, or to forward a request from Worker A to Worker B.

Service bindings provide the separation of concerns that microservice or service-oriented architectures provide, without configuration pain, performance overhead or knowledge of RPC protocols.

> - *Service bindings are fast.* When you use Service Bindings, communication between two Workers stays within Cloudflare. When one Worker invokes another, there is no network delay and the request is executed immediately.

Reviewer: Should we emphasize it typically runs literally on the same machine, in the same thread?

Otherwise this could be interpreted as "they're on a LAN together", which is how one would normally run microservices, so it sounds not that special.

Author: https://github.com/cloudflare/cloudflare-docs/pull/13252/commits/fb0da6750f14584de9e66b18e3681d2187c08e79

- *Service bindings are not just HTTP.* Worker A can expose methods that can be directly called by Worker B. Communicating between services only requires writing JavaScript methods and classes.

> 

Author: ```suggestion

![Service bindings are a zero-cost abstraction](/images/workers/platform/bindings/service-bindings-comparison.png)
```

Service bindings are commonly used to:

- *Provide a shared internal service to multiple Workers.* For example, you can deploy an authentication service as its own Worker, and then have any number of separate Workers communicate with it via Service bindings. 
- *Isolate services from the public Internet.* You can deploy a Worker that is not reachable via the public Internet, and can only be reached via an explicit Service binding that another Worker declares. 
- *Allow teams to deploy code independently.* Team A can deploy their Worker on their own release schedule, and Team B can deploy their Worker separately.

## Configuration

```toml
---
filename: wrangler.toml
---
services = [
  { binding = "<BINDING_NAME>", service = "<WORKER_NAME>" }
]
```

* `binding`: The name of the key you want to expose on the `env` object.
* `service`: The name of the target Worker you would like to communicate with. This Worker must be on your Cloudflare account.

## Interfaces

Worker A that declares a Service binding to Worker B can call Worker B in two different ways:

- *RPC* By calling any public method on the class that Worker B exports (`await env.BINDING_NAME.myMethod(arg1)`)

> - *HTTP* By calling the `fetch()` method on the binding (`env.BINDING_NAME.fetch(request)`)

Reviewer: Could maybe emphasize more that the name "fetch" is special, and can't be the name of an RPC method?

Author: https://github.com/cloudflare/cloudflare-docs/pull/13252/commits/432478089d741a2b686acbabd183a67a86d2f931

  async add(a, b) { return a + b; }
}
```

Which the following Worker declares a binding to:

```toml
---
filename: wrangler.toml
---
name = "worker_a"
main = "./src/workerA.js"
services = [
  { binding = "WORKER_B", service = "worker_b" }
]
```

And then calls as an async method:

```js
---
filename: workerA.js
---
export default {
  async fetch(request, env) {
    const result = await env.WORKER_B.add(1, 2);
    return new Response(result);
  }
}
```

This is a remote procedure call (RPC) protocol, without having to think about the protocol. The client, in this case Worker A, calls Worker B and tells it to execute a specific procedure using specific arguments that the client provides. But with Service bindings, there is no need to learn, implement or interact with a specific RPC protocol. It's just JavaScript.

#### Remote objects

Objects that Worker A receives back from Worker B can expose methods. These methods can be called from within Worker A. Consider the following example:

```js
---
filename: workerB.js
---
import { WorkerEntrypoint } from "cloudflare:workers";

export class WorkerB extends WorkerEntrypoint {
  async foo() {

>     return {

Reviewer: I wonder if we actually want to encourage this "return object literal" syntax vs. defining a class that inherits `RpcTarget`.

With an object literal, you're actually sending a bag of functions, where each function requires a separate stub. It's kind of inefficient. Also possibly confusing in that any non-function properties of the object are sent upfront and any server-side changes to them are not reflected client-side.

      bar: () => Math.random(),

>   }

Reviewer: Missing a closing brace here.

}
```

And then calls as an async method:

```js
---
filename: workerA.js
---
export default {
  async fetch(request, env) {
    const foo = await env.WORKER_B.foo();

>     const bar = await foo.bar();

Reviewer: We should maybe point out that `foo` needs to be disposed here. This can be done by calling `foo[Symbol.dispose]()`, or if we were using TypeScript, using `using`:

```
using foo = await env.WORKER_B.foo();
const bar = await foo.bar();
// foo is automatically disposed when it goes out-of-scope
```

Of course, in this particular example, disposal is not all that important because it'll happen implicitly at the end of the request context. But it becomes important pretty quickly when the logic gets more complicated.

```toml
---
filename: wrangler.toml
---
name = "worker_b"
main = "./src/workerB.js"
```

```js
---
filename: workerB.js
---
export default {
  async fetch(request, env, ctx) {
    return new Response("Hello World!");
  }
}
```

The following Worker declares a binding to the Worker above:

```toml
---
filename: wrangler.toml
---
name = "worker_a"
main = "./src/workerA.js"
services = [
  { binding = "WORKER_B", service = "worker_b" }
]
```

And then forwards a request to:

```js
---
filename: workerA.js
---
export default {
	async fetch(request, env) {
		return await env.BINDING_NAME.fetch(request);
	},
};
```

## Lifecycle

The Service bindings API is asynchronous — you must `await` any method you call. If Worker A invokes Worker B via a Service binding, and Worker A does not await the completion of Worker B, Worker B will be terminated early.

## Local development

Local development is supported for Service bindings. For each Worker, open a new terminal and use [`wrangler dev`](/workers/wrangler/commands/#dev) in the relevant directory or use the `SCRIPT` option to specify the relevant Worker's entrypoint.

> 

Author: TODO: Section about Smart Placement here, link to https://developers.cloudflare.com/workers/configuration/smart-placement/#best-practices

Author: ```suggestion
## Smart Placement

[Smart Placement](/workers/configuration/smart-placement) automatically places your Worker in an optimal location that minimizes latency.

You can use Smart Placement together with Service bindings to split your Worker into two services:

![Smart Placement and Service Bindings](/images/workers/platform/smart-placement-service-bindings.png)

Refer to the [docs on Smart Placement](/workers/configuration/smart-placement/#best-practices) for more.

```

Reviewer: Note that smart placement will not initially work for RPC (RPC requests will simply run the worker locally, without smart placement). We'll need to fix that later.

Author: https://github.com/cloudflare/cloudflare-docs/pull/13252/commits/0334d3accbf759cd7d061ebce7088f704577b3cf

## About Service bindings

[Service bindings](/workers/runtime-apis/bindings/service-bindings/) facilitate Worker-to-Worker communication. A Service binding allows Worker A to call a method on Worker B, or to forward a request from Worker A to Worker B.

Service bindings provide the separation of concerns that microservice or service-oriented architectures provide, without configuration pain, performance overhead or knowledge of RPC protocols.

- **Service bindings are fast.** When you use Service Bindings, communication between two Workers stays within Cloudflare. When one Worker invokes another, there is no network delay and the request is executed immediately.
- **Service bindings are not just HTTP.** Worker A can expose methods that can be directly called by Worker B. Communicating between services only requires writing JavaScript methods and classes.

![Service bindings are a zero-cost abstraction](/images/workers/platform/bindings/service-bindings-comparison.png)

> Service bindings are commonly used to:

Author: ```suggestion

Service bindings are commonly used to:
```

- *Isolate services from the public Internet.* You can deploy a Worker that is not reachable via the public Internet, and can only be reached via an explicit Service binding that another Worker declares.
- *Allow teams to deploy code independently.* Team A can deploy their Worker on their own release schedule, and Team B can deploy their Worker separately.

## Configuration

```toml
---
filename: wrangler.toml
---
services = [
  { binding = "<BINDING_NAME>", service = "<WORKER_NAME>" }
]
```

* `binding`: The name of the key you want to expose on the `env` object.
* `service`: The name of the target Worker you would like to communicate with. This Worker must be on your Cloudflare account.

## Interfaces

Worker A that declares a Service binding to Worker B can call Worker B in two different ways:

- *RPC* By calling any public method on the class that Worker B exports (`await env.BINDING_NAME.myMethod(arg1)`)
- *HTTP* By calling the `fetch()` method on the binding (`env.BINDING_NAME.fetch(request)`)

### RPC (`WorkerEntrypoint`)

> If you write your Worker as a JavaScript class that extends the built-in `WorkerEntryptoint` class, public methods on the class you export are exposed via Service bindings and can be called.

Reviewer 2: ```suggestion
If you write your Worker as a JavaScript class that extends the built-in `WorkerEntrypoint` class, public methods on the class you export are exposed via Service Bindings and can be called.
```

For example, the following Worker implements the public method `add(a, b)`:

```toml
---
filename: wrangler.toml
---
name = "worker_b"
main = "./src/workerB.js"
```

```js
---
filename: workerB.js
---
import { WorkerEntrypoint } from "cloudflare:workers";

export class WorkerB extends WorkerEntrypoint {
  async add(a, b) { return a + b; }
}
```

Which the following Worker declares a binding to:

```toml
---
filename: wrangler.toml
---
name = "worker_a"
main = "./src/workerA.js"
services = [
  { binding = "WORKER_B", service = "worker_b" }
]
```

And then calls as an async method:

```js
---
filename: workerA.js
---
export default {
  async fetch(request, env) {
    const result = await env.WORKER_B.add(1, 2);
    return new Response(result);
  }
}
```

> This is a remote procedure call (RPC) protocol, without having to think about the protocol. The client, in this case Worker A, calls Worker B and tells it to execute a specific procedure using specific arguments that the client provides. But with Service bindings, there is no need to learn, implement or interact with a specific RPC protocol. It's just JavaScript.

Reviewer 2: Should this be `Service Bindings` - title-case, as it's a noun?

Author: We don't capitalize it today:

https://developers.cloudflare.com/workers/configuration/bindings/

But seems like it could be capitalized across the board since it's a reference to a named concept on our platform, with specific meaning on Workers.

@deadlypants1973

## About Service bindings

[Service bindings](/workers/runtime-apis/bindings/service-bindings/) facilitate Worker-to-Worker communication. A Service binding allows Worker A to call a method on Worker B, or to forward a request from Worker A to Worker B.

> Service bindings provide the separation of concerns that microservice or service-oriented architectures provide, without configuration pain, performance overhead or knowledge of RPC protocols.

Author: ```suggestion
Service bindings provide the separation of concerns that microservice or service-oriented architectures provide, without configuration pain, performance overhead or need to learn RPC protocols.
```

- **Service bindings are fast.** When you use Service Bindings, communication between two Workers stays within Cloudflare. When one Worker invokes another, there is no network delay and the request is executed immediately.
- **Service bindings are not just HTTP.** Worker A can expose methods that can be directly called by Worker B. Communicating between services only requires writing JavaScript methods and classes.

![Service bindings are a zero-cost abstraction](/images/workers/platform/bindings/service-bindings-comparison.png)

Service bindings are commonly used to:

- **Provide a shared internal service to multiple Workers.** For example, you can deploy an authentication service as its own Worker, and then have any number of separate Workers communicate with it via Service bindings.
- **Isolate services from the public Internet.** You can deploy a Worker that is not reachable via the public Internet, and can only be reached via an explicit Service binding that another Worker declares.
- **Allow teams to deploy code independently.** Team A can deploy their Worker on their own release schedule, and Team B can deploy their Worker separately.

## Configuration

> 

Author: ```suggestion

You add a Service binding by modifying the `wrangler.toml` of the caller — the Worker that you want to be able to initiate requests.

For example, if you want Worker A to be able to call Worker B — you'd add the following to the `wrangler.toml` for Worker A:
```

]
```

And then forwards a request to:

```js
---
filename: workerA.js
---
export default {
	async fetch(request, env) {
		return await env.BINDING_NAME.fetch(request);
	},
};
```

## Lifecycle

The Service bindings API is asynchronous — you must `await` any method you call. If Worker A invokes Worker B via a Service binding, and Worker A does not await the completion of Worker B, Worker B will be terminated early.

## Local development

Local development is supported for Service bindings. For each Worker, open a new terminal and use [`wrangler dev`](/workers/wrangler/commands/#dev) in the relevant directory or use the `SCRIPT` option to specify the relevant Worker's entrypoint.
## Smart Placement

[Smart Placement](/workers/configuration/smart-placement) automatically places your Worker in an optimal location that minimizes latency.

You can use Smart Placement together with Service bindings to split your Worker into two services:

![Smart Placement and Service Bindings](/images/workers/platform/smart-placement-service-bindings.png)

Refer to the [docs on Smart Placement](/workers/configuration/smart-placement/#best-practices) for more.

> 

Author: Stubbing this out but needs more work:

```suggestion
### What types of objects can be passed to remote methods?

Only objects that are compatible with [`structuredClone()`](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) can be used when calling remote methods on another Worker.

For example:

- you cannot pass `Request` or `Response` objects to a method (other than the built-in `fetch()` handler)
```

Reviewer: We're actually going to support `Request` and `Response` (even though they aren't supported by structured clone generally).
